sos/plugins/__init__.py

```diff
@@ -146,34 +146,54 @@ def __repr__(self):
         """
         return self.__str(quote=True, prefix="SoSPredicate(", suffix=")")
 
-    def __nonzero__(self):
-        """Predicate evaluation hook.
-        """
-        pvalue = False
-
-        # Allow loading kernel modules?
-        pvalue |= self._owner.get_option("allow_system_changes")
+    def _eval_kmods(self):
+        if not self.kmods or self._owner.get_option('allow_system_changes'):
+            return True
 
         # Are kernel modules loaded?
-        for k in self.kmods:
-            pvalue |= self._owner.is_module_loaded(k)
+        _kmods = [self._owner.is_module_loaded(k) for k in self.kmods]
+
+        if self.required['kmods'] == 'any':
+            return any(_kmods)
+        else:
+            return all(_kmods)
 
-        for s in self.services:
-            pvalue |= self._owner.service_is_running(s)
+    def _eval_services(self):
+        if not self.services:
+            return True
+
+        if self.required['services'] == 'any':
```

Comment:
Can't be the logic from `_eval_kmods` used here as well? I.e.:

        _services = [self._owner.service_is_running(s) for s in self.services]

        if self.required['services'] == 'any':
            return any(_services)
        else:
            return all(_services)

That removes some code duplicity and imho is more readable code ("any"/"all").

Reply:
Yup, it totally should be.

Reply:
Updated the PR for this.